src/rdkafka_broker.c

```diff
@@ -5104,18 +5108,17 @@ static int rd_kafka_broker_thread_main (void *arg) {
                         /* Asynchronous connect in progress. */
                         rd_kafka_broker_serve(rkb, rd_kafka_max_block_ms);
 
-			if (rkb->rkb_state == RD_KAFKA_BROKER_STATE_DOWN) {
-				/* Connect failure.
-				 * Try the next resolve result until we've
-				 * tried them all, in which case we sleep a
-				 * short while to avoid busy looping. */
-				if (!rkb->rkb_rsal ||
-                                    rkb->rkb_rsal->rsal_cnt == 0 ||
-                                    rkb->rkb_rsal->rsal_curr + 1 ==
-                                    rkb->rkb_rsal->rsal_cnt)
-                                        rd_kafka_broker_serve(
-                                                rkb, rd_kafka_max_block_ms);
-			}
+                        /* Connect failure.
+                         * Try the next resolve result until we've
+                         * tried them all, in which case we back off the next
+                         * connection attempt to avoid busy looping. */
+			if (rkb->rkb_state == RD_KAFKA_BROKER_STATE_DOWN &&
```

Comment:
fwiw i like this written the old way, breaks it into steps. but inconsequential.

Reply:
I think abstracting the address roundrobin check to its own function will improve readability